### PR TITLE
[camera] android-rework part 1: Base classes to support Android Camera features

### PR DESCRIPTION
**Pull request by mvanbeusekom**
Introduces some base classes in the Android Camera implementation which will allow us to separate specific feature logic from the `Camera` class into specific feature classes. This has the following benefits (basically all benefits of the Separation of Concerns principle):

- It allows for focussed feature classes making it easier to maintain the implementation of the feature;
- It will reduce the amount of code currently in the `Camera` class, making it easier to maintain the `Camera` class;
- It will be easier to write unit-tests to cover the feature logic.

This is the first PR in a series of pull-requests which will gradually introduce changes from PR #3651, making it easier to review the code (as discussed with @stuartmorgan).

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
  - Since this will be a series of pull-requests I will update the version in the last PR of the series.
- [ ] I updated CHANGELOG.md to add a description of the change.
  - Since this will be a series of pull-requests I will update the version in the last PR of the series.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

**Comment by acoutts** [comment]
@googlebot I consent.

**Comment by stuartmorgan** [comment]
FYI I'm OOO at the moment but will start working through these reviews as time allows.

**Review by bparrishMines — APPROVED** [review]
LGTM

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraProperties.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+import android.hardware.camera2.CameraManager;
+import android.os.Build.VERSION_CODES;
+import android.util.Range;
+import android.util.Rational;
+import android.util.Size;
+import androidx.annotation.RequiresApi;
+
+/** An interface allowing access to the different characteristics of the device's camera. */
+public interface CameraProperties {
+  String getCameraName();
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraProperties.java:19]: It's not obvious what most of these methods mean. Please add Javadoc comments to each public method (as required by style); it seems like the best option would probably be to document them by indicating what CameraCharacteristics entry each one maps to.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraProperties.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  public Integer getControlMaxRegionsAutoExposure() {
+    return cameraCharacteristics.get(CameraCharacteristics.CONTROL_MAX_REGIONS_AE);
+  }
+
+  @Override
+  public Integer getControlMaxRegionsAutoFocus() {
+    return cameraCharacteristics.get(CameraCharacteristics.CONTROL_MAX_REGIONS_AF);
+  }
+
+  @RequiresApi(api = VERSION_CODES.P)
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraProperties.java:108]: Maybe this is a Java noob question, but shouldn't this be on the interface, not the implementation? It's not clear to me what it means for an interface with no API requirement to be implemented by something that does have one.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeature.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+  protected CameraFeature(@NonNull CameraProperties cameraProperties) {
+    this.cameraProperties = cameraProperties;
+  }
+
+  /** Debug name for this feature. */
+  public abstract String getDebugName();
+
+  /**
+   * Get the current value of this feature's setting.
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeature.java:30]: What is the expected behavior of get/set if `checkIsSupported` returns false? That would be good to document.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeature.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  /**
+   * Set a new value for this feature's setting.
+   *
+   * @param value
+   */
+  public abstract void setValue(T value);
+
+  /**
+   * Returns whether or not this feature is supported.
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeature.java:44]: Nit: please use consistent mood in declaration comments (e.g., the above is Sets, not Set, but this is Returns, not Return.)

It looks like the preferred style for Javadoc is declarative, so this is correct and the other methods here should be Gets, Sets, Updates.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Just small nits, otherwise looks good.

**Review by stuartmorgan — APPROVED** [review]
LGTM

**Pull request merged**